core/network/impl/peer_client_libp2p.cpp

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "network/impl/peer_client_libp2p.hpp"
+
+#include "common/buffer.hpp"
+#include "libp2p/basic/message_read_writer.hpp"
+#include "network/impl/common.hpp"
+#include "scale/scale.hpp"
+
+namespace kagome::network {
+  using common::Buffer;
+  using libp2p::basic::MessageReadWriter;
+
+  PeerClientLibp2p::PeerClientLibp2p(libp2p::Host &host,
+                                     libp2p::peer::PeerInfo peer_info,
+                                     common::Logger logger)
+      : host_{host},
+        peer_info_{std::move(peer_info)},
+        log_{std::move(logger)} {}
+
+  void PeerClientLibp2p::blocksRequest(BlocksRequest request,
+                                       BlockResponseHandler handler) const {
+    // request-response model requires us to open "sync" channel
+    host_.newStream(
+        peer_info_,
+        kSyncProtocol,
+        [self{shared_from_this()},
+         request{std::move(request)},
+         cb{std::move(handler)}](auto &&stream_res) mutable {
+          if (!stream_res) {
+            self->log_->error("cannot open a stream to peer {}",
+                              self->peer_info_.id.toBase58());
+            return cb(stream_res.error());
+          }
+          auto stream = std::move(stream_res.value());
+
+          auto encoded_request_res = scale::encode(request);
+          if (!encoded_request_res) {
+            self->log_->error(
+                "cannot open a encode a block request message: {}",
+                encoded_request_res.error().message());
+            return cb(encoded_request_res.error());
+          }
+          auto encoded_request =
+              std::make_shared<Buffer>(std::move(encoded_request_res.value()));
+
+          auto read_writer =
+              std::make_shared<MessageReadWriter>(std::move(stream));
+          read_writer->write(
```

Comment:
why do you answer with REQUEST to a user when you need to answer with RESPONSE?

Reply:
This is a client and it makes a `request`, right?

Reply:
Solved the issue with @Warchant one-to-one